pkg/syncer/spec/spec_process.go

```diff
@@ -114,14 +115,30 @@ func (c *Controller) process(ctx context.Context, gvr schema.GroupVersionResourc
 	}
 	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
 
-	// to downstream
-	downstreamNamespace, err := shared.PhysicalClusterNamespaceName(shared.NamespaceLocator{
-		LogicalCluster: clusterName,
-		Namespace:      upstreamNamespace,
-	})
-	if err != nil {
-		klog.Errorf("Error hashing namespace %s|%s: %v", clusterName, upstreamNamespace, err)
-		return nil // ignore error, shouldn't happen
+	namespaceGvr := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "namespaces"}
+	desiredNSLocator := shared.NewNamespaceLocator(clusterName, c.workloadClusterLogicalClusterName, c.workloadClusterUID, c.workloadClusterName, upstreamNamespace)
+	downstreamNamespaces := c.downstreamInformers.ForResource(namespaceGvr).Informer().GetIndexer().List()
+
+	var downstreamNamespace string
+	for _, namespace := range downstreamNamespaces {
```

Comment:
now you are using the indexer as a non-index. The index can give you the correct namespace in constant time. See indexer.ByIndex.

Reply:
Yes... the idea is for the Spec Syncer to be able to figure out the namespace without having to calculate the hash, this way we will be able to change the pattern and just rely on the downstream namespaceLocator, but yes, we need to lookup between all the namespaces. 

This could be improved by checking if the namespace from the hashed namespaceLocator exists (physical...), and then just use that one, if not, look for it on the list of namespaces and try to find another namespace with the desired namespaceLocator.

Reply:
we cannot look at a list of namespaces. We have to use an index on namespaces here. Not sure we talk about the same.